File: companion/bolumler/eksik-veri/ornek-01/cozum.py
import argparse
from pathlib import Path
import numpy as np
import pandas as pd
from scipy import stats

parser = argparse.ArgumentParser()
parser.add_argument('--check', action='store_true')
args = parser.parse_args()
d = pd.read_csv('veri.csv')
assert list(d.columns) == ['grup','on','son']
assert d.grup.isin([1,2]).all() and d['on'].between(0,100).all()
assert d.son.dropna().between(0,100).all()
s = {'n':len(d), 'eksik_son':d.son.isna().sum(), 'eksik_oran':d.son.isna().mean()}
for g in [1,2]:
    x=d.loc[d.grup==g,'son'].dropna()
    assert len(x)>1
    s.update({f'n_{g}':len(x),f'ortalama_{g}':x.mean(),f'ss_{g}':x.std(ddof=1)})
s['on_eksik_son'] = d.loc[d.son.isna(),'on'].mean()
s['on_gozlenen_son'] = d.loc[d.son.notna(),'on'].mean()
out=pd.DataFrame(list(s.items()),columns=['olcut','deger'])
print(out.to_string(index=False))
out.to_csv('sonuclar-python.csv',index=False)
if args.check:
    ref=pd.read_csv('beklenen-sonuclar.csv')
    if list(ref.olcut)!=list(out.olcut): raise AssertionError('Sonuç etiketleri farklı.')
    np.testing.assert_allclose(out.deger,ref.deger,rtol=1e-8,atol=1e-10)
    print('KONTROL BAŞARILI')

File: companion/bolumler/eksik-veri/ornek-01/cozum.R
d <- read.csv("veri.csv", na.strings=c("", "NA"))
stopifnot(identical(names(d),c("grup","on","son")), all(d$grup %in% 1:2), all(d$on>=0 & d$on<=100), all(d$son>=0 & d$son<=100,na.rm=TRUE))
s <- c(n=nrow(d),eksik_son=sum(is.na(d$son)),eksik_oran=mean(is.na(d$son)))
for(g in 1:2){x<-d$son[d$grup==g];x<-x[!is.na(x)];stopifnot(length(x)>1); s<-c(s,setNames(c(length(x),mean(x),sd(x)),paste0(c("n_","ortalama_","ss_"),g)))}
s<-c(s,on_eksik_son=mean(d$on[is.na(d$son)]),on_gozlenen_son=mean(d$on[!is.na(d$son)]))
out<-data.frame(olcut=names(s),deger=as.numeric(s))
print(out,row.names=FALSE);write.csv(out,"sonuclar-r.csv",row.names=FALSE)
if("--check" %in% commandArgs(trailingOnly=TRUE)){
 ref<-read.csv("beklenen-sonuclar.csv")
 stopifnot(identical(out$olcut,ref$olcut),all(abs(out$deger-ref$deger)<=1e-10+1e-8*abs(ref$deger)))
 cat("KONTROL BASARILI\n")
}
sessionInfo()
